python/pyspark/exceptions.py

```diff
@@ -0,0 +1,42 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+class NonDriverReferenceError(Exception):
+    """
+    Thrown when trying to reference a driver's value or attribute in a non driver environment.
+    """
+
+
+class JavaGatewayError(Exception):
+    """
+        Indicates an error of the java gateway.
+    """
+
+
+class SparkContextInitializationError(RuntimeError):
```

Comment:
Hmm, do we need a complex exception hierarchy like this? There are exceptions too specific, e.g. SparkContextInitializationError, SparkContextConfigurationError. I think even in Scala codebase we don't throw specific exception for configuration error.

If we follow this direction, I'm afraid we will have more and more exception classes.

Reply:
Yes I can see your point about the complex hierarchy and I've committed a revision that merges both of the Spark Context's exceptions. With that said I think that the relative complexity and possibilities of errors thrown when initializing a spark context does merit an exception class of its own

Reply:
For example, you can see the immediate benefit in wrapping java gateway errors with this exception type [here](https://github.com/apache/spark/pull/31238/files#diff-1d250699e56249e8bc181a6c531817f4aed14120fe1daf04ced83dc947dddc0aR336)

Reply:
Just my two cents ‒ while adding more specific exception classes can be useful, especially in contexts where we might expect user to recover, there is little point in building hierarchy for types of conditions which are effectively a programming mistake.

For example `NonDriverReferenceError` refers to condition, from which there is no meaningful way out. Such cases should rather provide human readable description of the problem, so user can easily identify and fix the issue (@HyukjinKwon  and others already did a lot of work in that direction). Similarly `JavaGatewayErrors` are typically fatal and there is little that user can do there, without tinkering with internals.

Reply:
Yeah .. I echo to @viirya and @zero323. We can start with simple ones with a good description.